lib/graphql/input_object_type.rb

```diff
@@ -147,6 +153,11 @@ def validate_non_null_input(input, ctx)
         end
       end
 
+      # Better way to do this, perhaps keeping the coerced value around?
+      if coerce_non_null_input(input, ctx).nil?
```

Comment:
Introduces a performance regression: input is converted into the `InputObject` class created by the user, and the custom `prepare` method is invoked, but all of that work is thrown away.

If the user's `initialize` and `prepare` methods are expensive, it'll cause a problem. Also, they might not expect those methods to get invoked twice: once for validation, and then once for the actual value.